build/dart/rules.gni

```diff
@@ -5,7 +5,178 @@
 # This file has rules for making Dart packages and Dart-based Mojo applications.
 # The entrypoint is the dart_pkg rule.
 
+import("//build/compiled_action.gni")
 import("//build/module_args/dart.gni")
+import("//flutter/common/config.gni")
+import("//third_party/dart/build/dart/dart_action.gni")
+
+# Creates a dart kernel (dill) file suitalbe for use with gen_snapshot.
+#
+# Invoker must supply dart_main and package_config. Invoker may optionally
+# supply aot as a boolean and product as a boolean.
+template("dart_snapshot") {
+  assert(defined(invoker.main_dart), "main_dart is a required parameter.")
+  assert(defined(invoker.sources), "sources is a required parameter.")
+  assert(defined(invoker.package_config),
+         "package_config is a required parameter.")
+
+  kernel_target = "_${target_name}_kernel"
+  snapshot_target = "_${target_name}_snapshot"
+  is_aot =
+      flutter_runtime_mode == "profile" || flutter_runtime_mode == "release"
+
+  kernel_output = "$target_gen_dir/kernel_blob.bin"
+
+  prebuilt_dart_action(kernel_target) {
+    main_dart = rebase_path(invoker.main_dart)
+    package_config = rebase_path(invoker.package_config)
+
+    platform_dill = "$root_out_dir/flutter_patched_sdk/platform_strong.dill"
```

Comment:
Minor nit: Presumably, these are outputs of the `//flutter/lib/snapshot` dep below. Can we use `get_label_info` with `target_gen_dir` to determine the location of these artifacts? Right now, these paths are hardcoded in unrelated rules.

Reply:
This isn't in the `target_gen_dir` of that target - it's hard coded in there. 

Changing that is possible, but right now recipes depend on this location being where it is.

Reply:
I have a goal of cleaning up the way recipes get artifacts though - and once that's all set, we can start changing these things more easily.